Refactor and extend NAT with UDP support

TMVector opened the pull request:
This PR is a neatened and improved version of [#9](https://github.com/niksu/pax/pull/9).

Refactor the NAT to reflect the approach of an applications software engineer. Also add support for UDP, timeouts, TCP state, and TIME_WAIT.

niksu commented:
Thanks for tidying up the PR, I'll make another pass through the code to give any remaining feedback. I'm interested in the workflow you followed -- how did you go about tidying it up? Did you fuse commits together and/or make new ones, or rearranged existing ones? Was cherrypick sufficiently good for your purposes or did you use a different tool?

niksu commented:
Another workflow-related question: how did you go about testing the features added in this PR, to check that UDP, timeouts, and TCP state are handled well? Did you use a small set of system tests, or did you use fuzzing to test a larger space of parameters? Thanks!

TMVector commented:
I cherrypicked the commits that were okay from the first PR, but since the commits weren't very granular, I thought it best to focus on fixing the problems than isolating changes into separate commits. I checked out the working directory from #9 and made my changes, and pretty much committed changes as I normally would.

I'm afraid testing was pretty limited. I used the tests that are in the python scripts, and while I was creating the tests I did a fair bit of manually playing around.

niksu commented:
I did a quick pass so far and thought I'd send some high-level feedback. I'll do a closer pass soon. The code looks in great shape. On a pedantic note, I noticed that you removed trailing whitespace :) Please also ensure that there's a single space between `def` blocks in the Python code, and add the standard header to files that lack it, such as `examples/Nat/TcpNAT.cs` and `examples/Nat/UdpNAT.cs`.

Seeing that the classes in `Nat\` have proliferated it would be very helpful to have a README.md file there that briefly describes the role of each class and how it relates to others (e.g., "class X uses class Y", "class X extends class Y", "class X contains class Y because...", etc), so one can more easily obtain an idea of the overall class structure.

The testing infrastructure in the Python script looks very sophisticated, it's great practice that you're developing that together with the code it tests! I'd say that feature-wise the NAT is at a great place. Thanks for restructuring it, extending it, and taking care of subtle cases to do with timers etc.

niksu commented:
Quick question, could you confirm if all the points that were raised during the discussion of https://github.com/niksu/pax/pull/9 were addressed in the current PR? Thanks!

niksu commented on examples/Nat/NAT.cs:
```diff
@@ -0,0 +1,144 @@
+      }
+
+      // If we reach this point then we can't handle that type of packet
+      return new ForwardingDecision.SinglePortForward(Port_Drop);
```
Have you considered simply using `ForwardingDecision.Drop` here?

TMVector replied:
Unfortunately I can't because it isn't of type `SinglePortForward`

niksu commented on examples/Nat/NAT.cs:
```diff
@@ -0,0 +1,144 @@
+    /// <param name="incomingNetworkInterface">The number representing network interface the packet arrived on.</param>
+    /// <param name="packet">The observed packet.</param>
+    /// <returns>A <see cref="ForwardingDecision"/> for the packet.</returns>
+    override public ForwardingDecision process_packet(int incomingNetworkInterface, ref Packet packet)
```
Nice, I like how this function makes it easy for some core logic to bubble up to the surface of the code, and protocol-specific handling (e.g., for TCP) is done in dedicated handlers.

niksu commented on examples/Nat/NATBase.cs:
```diff
@@ -0,0 +1,287 @@
+  /// <typeparam name="TPacket">The type of packet.</typeparam>
+  /// <typeparam name="TNode">The type of node.</typeparam>
+  /// <typeparam name="TEncapsulation">The type of the packet encapsulation.</typeparam>
+  public abstract class NATBase<TPacket,TNode,TEncapsulation> where TPacket : Packet where TNode : Node where TEncapsulation : PacketEncapsulation<TPacket,TNode>
```
I think it's very elegant how this class factors out the NAT behaviour that can be instantiated by different protocols.

niksu commented on examples/Nat/NATBase.cs:
```diff
@@ -0,0 +1,287 @@
+    /// <param name="ipAddress">The IP address to use.</param>
+    /// <param name="interfaceNumber">The interface number to use.</param>
+    /// <param name="macAddress">The MAC address to use.</param>
+    protected abstract TNode CreateMasqueradeNode(IPAddress ipAddress, int interfaceNumber, PhysicalAddress macAddress);
```
I noticed that `TNode` is being used both as a name of an abstract class, and as a generic parameter (e.g., in `NatConnection`). It would be preferable to avoid this shadowing.

Also it's not immediately clear what the purpose of `TNode` is -- could you elaborate a bit in the description please?

niksu replied:
Sorry, `TNode` only seems to be used as a generic parameter, usually for subtypes of `Node`. It's still not clear to me how this function fits in the overall sequence of actions performed by the NAT. A brief high-level description should clear this up. Thanks!

TMVector replied:
I have updated the description of this method. Please let me know if you think it needs to be clearer or if you have any thoughts on a better name for the method.

niksu commented on examples/Nat/NatConnection.cs:
```diff
@@ -0,0 +1,80 @@
+    public void ReceivedPacket(PacketEncapsulation<TPacket,TNode> packet, bool packetFromInside)
+    {
+      // Mark used
+      DateTime used = DateTime.Now;
```
How precise is this timestamp? (e.g., microseconds? Would be useful to add this as a comment.)

niksu commented on examples/Nat/PacketEncapsulation.cs:
```diff
@@ -0,0 +1,213 @@
+    /// </summary>
+    /// <param name="incomingNetworkInterface">The network interface this packet arrived on.</param>
+    /// <returns>A <see cref="TNode"/> representing the source of the packet.</returns>
+    public abstract TNode GetSourceNode(int incomingNetworkInterface = -1);
```
Is `-1` intended to be a rogue value? If not, what does it mean? (Worth adding to the `param` comment if `incomingNetworkInterface` is expected to be nonnegative.)

niksu commented on examples/Nat/TcpNAT.cs:
```diff
@@ -0,0 +1,72 @@
+    /// <summary>
+    /// The end of the range of TCP ports to use (inclusive).
+    /// </summary>
+    private readonly ushort EndPort; // FIXME we don't use EndPort yet
```
Since `EndPort` isn't being used, am I correct in thinking that currently (in theory) assigned port numbers could reach the maximum value of `ushort`?

TMVector replied:
Yes. It would then become `0` and wrap around to `StartPort`. I had planned to implement a range constraint, but it doesn't remove the problem of wrapping around and reusing port numbers that may still be in use.

niksu replied:
It would be useful to extend this comment to warn of that possible behaviour.

TMVector commented on examples/Nat/TcpState.cs:
```diff
@@ -0,0 +1,126 @@
+      TIME_WAIT = time_wait;
+    }
+
+    public void UpdateState(TcpPacket packet, bool packetFromInside)
```
> What happens if we receive a repeated Syn travelling in the same direction? (From inside to out, say) Is the machine reset? That is, can evil person (call them E) break user A's connection to outside server B, when E sends the NAT a Syn addressed to B (and spoofed to appear to come from A)?

I'm not sure if I've changed the code since this comment was made. A malicious Syn would be ignored, but a malicious Fin would cause the NAT entry to be removed after the TIME_WAIT timeout was finished regardless of continuing traffic. It would also be easier than normal spoofing because the attacker doesn't need to know the sequence number for the packet to be accepted by the NAT as valid.

niksu replied:
Thanks! It would be very useful to have this documented as a comment, 
since it exposes the limitations of the system and flags up possible 
improvements.

On 2016-07-18 15:11, TMVector wrote:
> In examples/Nat/TcpState.cs [1]:
> 
>> +
>> + /// <summary>
>> + /// True if the TCP connections in both directions are closed,
> determining if the connection can be removed before the timeout
> elapses.
>> + /// </summary>
>> + public bool CanBeClosed { get { return ClosedFromInside &&
> ClosedFromOutside && !InTimeWait; } }
>> +
>> + /// <summary>
>> + /// Creates a new TcpState in the initial state.
>> + /// </summary>
>> + /// <param name="time_wait">The duration of the TIME_WAIT
> state.</param>
>> + public TcpState(TimeSpan time_wait)
>> + {
>> + TIME_WAIT = time_wait;
>> + }
>> +
>> + public void UpdateState(TcpPacket packet, bool packetFromInside)
> 
>> What happens if we receive a repeated Syn travelling in the same
>> direction? (From inside to out, say) Is the machine reset? That is,
>> can evil person (call them E) break user A's connection to outside
>> server B, when E sends the NAT a Syn addressed to B (and spoofed to
>> appear to come from A)?
> 
> I'm not sure if I've changed the code since this comment was made. A
> malicious Syn would be ignored, but a malicious Fin would cause the
> NAT entry to be removed after the TIME_WAIT timeout was finished
> regardless of continuing traffic. It would also be easier than normal
> spoofing because the attacker doesn't need to know the sequence number
> for the packet to be accepted by the NAT as valid.
> 
> --
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1] https://github.com/niksu/pax/pull/12#discussion_r71153426
> [2]
> https://github.com/niksu/pax/pull/12/files/8f1fd2f3c9e5b30ce2ef0bb72ff888a4a096c87c#r71153426
> [3]
> https://github.com/notifications/unsubscribe-auth/ABDTTgrF3v9d9yjEsq9dgcJUOikTzWOjks5qW4l2gaJpZM4JNj7N

-- 
http://www.cl.cam.ac.uk/~ns441/

niksu replied:
This function is quote important, and deserves a couple of sentences of comment describing its intended behaviour. It's not immediately obvious from the code how the state is being transitioned and what role the parameters play.

TMVector commented:
I think I have addressed all your feedback from this and the last PR. Could you please check the description of [this](https://github.com/niksu/pax/pull/12/files#diff-3465279ee2070c9b5f37aefb4a0d1cfcR224) method - I have updated it, and I want to make sure it's clear what it is for from the comments in the code.
Thanks :smiley:

niksu commented:
Since the NAT is in its own directory under [examples](https://github.com/TMVector/pax/tree/fc467b965b85a5ff61b48d3f95592a9cbfc1856b/examples), it seems suitable to move nat_* files under that directory too. Currently this set consists of `nat_wiring.son`, `nat_topo.py`, and `nat_scapy_tests.py`.

niksu commented:
On your setup does the mininet test show `WARNING scapy test #1 failed. client 126, server 126` and `WARNING scapy test #2 failed. client 126, server 126` too, in output from the following command?

```
$ sudo ./examples/nat_topo.py --no-X test
action: test, X_windows: False, hold_open: False
*** Creating network
*** Adding controller
*** Adding hosts:
in1 in2 nat0 out0 
*** Adding switches:

*** Adding links:
(nat0, in1) (nat0, in2) (nat0, out0) 
*** Configuring hosts
in1 in2 nat0 out0 
*** Starting controller
c0 
*** Starting 0 switches

Network started
Setting the gateway on in1 to nat0
Setting the gateway on in2 to nat0
Starting Pax NAT process on nat0:
  nat0> $ Bin/Pax.exe examples/nat_wiring.json examples/Bin/Examples.dll
Connecting from in1 to out0:
  out0> $ echo Hello, are you there? | netcat -l 12001 &
  in1> $ netcat -n 10.0.0.4 12001
  in1> RCV: 'Hello, are you there?'
Correct data received

Sending data via UDP from in1 to out0:
  out0> $ echo Yes, UDP | nc -u -l 12001 -q1
  in1> $ echo Hello, are you there UDP? | nc -u 10.0.0.4 12001 -q1
  out0> RCV: 'Hello, are you there UDP?'
  in1> RCV: 'Yes, UDP'
Correct data received on the server
Correct data received on the client

Scapy test #1
  out0> $ examples/nat_scapy_tests.py server 35001
  in1> $ sleep 1
  in1> $ examples/nat_scapy_tests.py client
  in1> $ echo $?
  out0> $ echo $?
WARNING scapy test #1 failed. client 126, server 126

Scapy test #2
  out0> $ examples/nat_scapy_tests.py server2 35002
  in1> $ sleep 1
  in1> $ examples/nat_scapy_tests.py client2
  in1> $ echo $?
  out0> $ echo $?
WARNING scapy test #2 failed. client 126, server 126
Show Pax output? (y/N)
```

niksu commented:
I'm not sure if this is a more general issue, but mininet seems to break when I leave the CLI. I load the `run()` function as follows:

```
$ sudo ./examples/nat_topo.py --no-X run
action: run, X_windows: False, hold_open: False
*** Creating network
*** Adding controller
*** Adding hosts:
in1 in2 nat0 out0 
*** Adding switches:

*** Adding links:
(nat0, in1) (nat0, in2) (nat0, out0) 
*** Configuring hosts
in1 in2 nat0 out0 
*** Starting controller
c0 
*** Starting 0 switches

Network started
Setting the gateway on in1 to nat0
Setting the gateway on in2 to nat0
*** Starting CLI:
mininet> 
```
If I then press `^D` I get:

```
*** Stopping 1 controllers
c0 
*** Stopping 3 links
...
*** Stopping 0 switches

*** Stopping 4 hosts
in1 in2 nat0 out0 
*** Done
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "build/bdist.linux-x86_64/egg/mininet/cli.py", line 87, in <lambda>
    atexit.register( lambda: write_history_file( history_path ) )
IOError: [Errno 13] Permission denied
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "build/bdist.linux-x86_64/egg/mininet/cli.py", line 87, in <lambda>
IOError: [Errno 13] Permission denied
```

Do you get similar behaviour at your end?

niksu commented on examples/Nat/ConnectionKey.cs:
```diff
@@ -0,0 +1,67 @@
+
+    public override int GetHashCode() { return hashCode; }
+
+    // FIXME should all ToString overrides be conditional on target? (i.e. not on Kiwi)
```
Could you elaborate a bit on this thought please? Indeed the type `string` might not be possible or meaningful (in the same way) on some backends, and if `ToString()` is called purely for diagnostic/debugging reasons then this behaviour could be removed when targeting Kiwi say.

niksu commented on examples/Nat/ITransportState.cs:
```diff
@@ -0,0 +1,60 @@
+  public interface ITransportState<T> where T : Packet
+  {
+    /// <summary>
+    /// Gets a boolean value determining if the connection can be removed before the timeout elapses.
```
I think it needs to made clearer what role this can play. I think it's meant to capture the situation, as one might find in TCP, where a connection has been closed (via FIN exchange) and therefore we don't need to wait until the records expiry for the GC to remove the entry. But then even in TCP, doesn't the "TIME-WAIT" timer keep a connection from being considered "closed" until the timer expires? It's not clear what "timeout" refers to in the comment.

niksu commented on examples/NAT.cs:
```diff
@@ -1,379 +0,0 @@
-/*
```
Maybe it's worth keeping the old implementation, renaming it to something like "Simple NAT" perhaps, since it has the nice quality of being succinct, and serves as an example that one can get useful behaviour using relatively little code. (This also serves as the [example in the front README](https://github.com/niksu/pax#what-does-a-pax-processor-look-like).)

niksu replied:
On second thoughts, and having just read more carefully the implementation of the `NATBase` class, we could update the example in the front README to point to the functions [handlePacket](https://github.com/niksu/pax/pull/12/files#diff-3465279ee2070c9b5f37aefb4a0d1cfcR80) and [OutsideToInside](https://github.com/niksu/pax/pull/12/files#diff-3465279ee2070c9b5f37aefb4a0d1cfcR96) which succinctly give the core logic in the much-improved version of the NAT you developed.

niksu commented on examples/Nat/NAT.cs:
```diff
@@ -0,0 +1,148 @@
+    /// <param name="udp_start_port">The start of the range of UDP ports to use (inclusive).</param>
+    /// <param name="udp_end_port">The end of the range of UDP ports to use (inclusive).</param>
+    /// <remarks>
+    /// The constructor parameters can be specified in the wiring configuration file, in the `args` dictionary of the handler entry.
```
I very much like how you arranged for these parameters to be cleanly (and type-safely) extracted from the config (and allowed for optional parameters, and the support for multiple constructors in a clean way).

niksu commented on examples/Nat/NAT.cs:
```diff
@@ -0,0 +1,148 @@
+      TimeSpan tcp_inactivity_timeout, TimeSpan tcp_time_wait_duration, ushort tcp_start_port, ushort tcp_end_port,
+      TimeSpan udp_inactivity_timeout, ushort udp_start_port, ushort udp_end_port)
+    {
+      // Instantiate each NAT specialisation
```
This function looks very tidy! It calls protocol-specific implementations of the NAT.

niksu commented on examples/Nat/PacketEncapsulation.cs:
```diff
@@ -0,0 +1,213 @@
+    public abstract TNode GetDestinationNode(int incomingNetworkInterface = NAT.Port_Drop);
+
+    /// <summary>
+    /// This method is called whenever the SetSource method is called.
```
It might be useful to have a short example explaining a situation when `SetSource` is called -- is it called frequently and in different circumstances, or rarely and in similar circumstances?

niksu commented on examples/Nat/TcpState.cs:
```diff
@@ -0,0 +1,130 @@
+        state = TcpDirectionalState.Syn;
+    }
+
+    private void TransitionState(ref TcpDirectionalState state, bool ack)
```
`ack` parameter is not used in this function's body

TMVector replied:
Good spot!

niksu commented on examples/Nat/UdpNAT.cs:
```diff
@@ -0,0 +1,77 @@
+    /// <summary>
+    /// The end of the range of UDP ports to use (inclusive).
+    /// </summary>
+    private readonly ushort EndPort; // FIXME we don't use EndPort yet
```
Could reference the similar behaviour in the `TcpNAT` class here.

niksu commented on examples/nat_scapy_tests.py:
```diff
@@ -0,0 +1,265 @@
+
+    # Try sending data - should fail
+    print "Trying to send a packet to the client - should fail"
+    send(ip/TCP(sport=serverport, dport=natport, flags="", options=[('MSS', 1460)])/"FAIL PLEASE")
```
How do we notice if it has failed or succeeded?

niksu commented on examples/nat_topo.py:
```diff
@@ -134,40 +197,105 @@ def test():
+
+        # Create the pipe
+        pipe = "/tmp/cmdpipe%d" % (random.randint(0,1000))
+        h.cmd("mkfifo %s" % pipe)
```
Is the fifo removed after it's no longer needed?

niksu commented on examples/nat_scapy_tests.py:
```diff
@@ -0,0 +1,265 @@
+    # Establish connection to the server
+    tcp_syn = TCP(sport=clientport, dport=serverport, flags="S", options=[('MSS', 1460)])
+    print "Sending syn"
+    synack = sr1(ip/tcp_syn)[0]
```
It would be prudent here to check that the Syn and Ack flags in the received packet are indeed set.

niksu commented on examples/nat_scapy_tests.py:
```diff
@@ -0,0 +1,265 @@
+
+    # Wait for the connection to be initialised from the client
+    print "Waiting for connection from client"
+    syn = sniff(count=1, filter=filter)[0]
```
It would be prudent here to check that the Syn flag in the received packet is indeed set.

niksu commented on examples/nat_scapy_tests.py:
```diff
@@ -0,0 +1,265 @@
+    else:
+        return 0
+
+def run_server2(serverport=12022, natport=35002):
```
How does `run_server2` differ from `run_server`?

niksu commented on examples/nat_scapy_tests.py:
```diff
@@ -0,0 +1,265 @@
+    ## Set up and tear down connection
+    # Wait for Syn
+    print "Waiting for Syn from client"
+    syn = sniff(count=1, filter=filter)
```
As in `run_server2`, would be best to check that the packet we got has its Syn flag set, to validate our assumption.

niksu commented on examples/nat_scapy_tests.py:
```diff
@@ -0,0 +1,265 @@
+    else:
+        return 0
+
+def run_client2(serverport=12022, clientport=12021):
```
How does this differ from `run_client`?

niksu commented:
Great, thanks! I think it's much clearer.


On 2016-07-18 18:25, TMVector wrote:
> I think I have addressed all your feedback from this and the last PR.
> Could you please check the description of this [1] method - I have
> updated it, and I want to make sure it's clear what it is for from the
> comments in the code.
>  Thanks 😃
> 
> --
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1]
> https://github.com/niksu/pax/pull/12/files#diff-3465279ee2070c9b5f37aefb4a0d1cfcR224
> [2] https://github.com/niksu/pax/pull/12#issuecomment-233397561
> [3]
> https://github.com/notifications/unsubscribe-auth/ABDTTlLJyRzU8p9u5IrF40yxt_W4ERCSks5qW7b-gaJpZM4JNj7N

-- 
http://www.cl.cam.ac.uk/~ns441/

TMVector commented:
Regarding the Mininet/testing errors, I have not seen either before. I have fixed the shebang in the python scripts in case it helps with the 126 exit codes, but I don't know what would cause that. According to [Google](http://tldp.org/LDP/abs/html/exitcodes.html), 126 normally means that the command isn't executable, like `/dev/null`, so it could be a permissions issue?

I just tried testing over SSH from my laptop, and it worked okay, but I had to run `chmod a+x examples/Nat/*.py` first. I'm not sure how to fix that permanently though.

Aside from that, I think I have addressed all of your feedback - thanks very much for it :)

There are still a couple of things I would like to fix at some point:
- The scapy tests are too fragile
- The output of the scapy tests is only visible in the opened terminal windows, so you can't see it when using `--no-X`, or (vitally) if the error causes the terminal to close.

niksu commented on examples/Nat/TcpState.cs:
```diff
@@ -105,12 +105,15 @@ private void TransitionState(ref TcpDirectionalState state, bool syn, bool fin)
-      {
-        CloseTime = DateTime.Now;
-        state = TcpDirectionalState.FinAck;
+      if (ack)
```
I think I lost the flow of this function and how it fits in the big picture -- why must `ack==true` for the function to have any effect? Perhaps add a comment to the function to sketch its role wrt ack? Thanks

niksu commented on examples/Nat/nat_topo.py:
```diff
@@ -113,7 +113,7 @@ def test():
     # Start it in a separate terminal so that we can see the output in real time.
     print "Starting Pax NAT process on %s:" % nat0
-    cmd = 'Bin/Pax.exe examples/Nat/nat_wiring.json examples/Bin/Examples.dll'
+    cmd = 'Bin/Pax.exe examples/Nat/nat_wiring_test.json examples/Bin/Examples.dll'
```
Out of interest are the two configurations intended to differ other than by timeouts?

TMVector replied:
No, it's just so tests can be completed quicker really






On Tue, Jul 26, 2016 at 11:39 PM +0100, "Nik Sultana" <notifications@github.com> wrote:





> @@ -113,7 +113,7 @@ def test():
>      # Start the Pax NAT process on the NAT node:
>      # Start it in a separate terminal so that we can see the output in real time.
>      print "Starting Pax NAT process on %s:" % nat0
> -    cmd = 'Bin/Pax.exe examples/Nat/nat_wiring.json examples/Bin/Examples.dll'
> +    cmd = 'Bin/Pax.exe examples/Nat/nat_wiring_test.json examples/Bin/Examples.dll'

Out of interest are the two configurations intended to differ other than by timeouts?

---
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub:
https://github.com/niksu/pax/pull/12/files/07277e9a85e90c2258009c9fc280847c8a1c0268..ef7115111140c1f4b02c2219b8bfe9cde974bc38#r72351361

niksu commented:
Thanks, I'll retry running the commands involved in comments 
https://github.com/niksu/pax/pull/12#issuecomment-234261042 and 
https://github.com/niksu/pax/pull/12#issuecomment-234263211 and get back 
to you.

Regarding the fragility of the tests, perhaps take a look at the P4Paxos 
setup at 
https://github.com/usi-systems/p4paxos-demo/tree/master/p4paxos/bmv2 to 
enable us to test the Pax implementation of Paxos on that setup.

I think `chmod a+x examples/Nat/*.py` can be "committed" to the repo -- 
IIRC git also tracks rudimentary file permissions.

Finally, could you merge the current master branch this this PR's branch 
so the PR can be automatically mergeable via GitHub? I think the merge 
should be minor, probably involving a tiny conflict on Examples.csproj. 
Thanks again for your sustained effort on this, it's a great addition to 
the functionality and usability of Pax!

On 2016-07-26 15:01, TMVector wrote:
> Regarding the Mininet/testing errors, I have not seen either before. I
> have fixed the shebang in the python scripts in case it helps with the
> 126 exit codes, but I don't know what would cause that. According to
> Google [1], 126 normally means that the command isn't executable, like
> /dev/null, so it could be a permissions issue?
> 
> I just tried testing over SSH from my laptop, and it worked okay, but
> I had to run chmod a+x examples/Nat/*.py first. I'm not sure how to
> fix that permanently though.
> 
> Aside from that, I think I have addressed all of your feedback -
> thanks very much for it :)
> 
> There are still a couple of things I would like to fix at some point:
> 
>  	* The scapy tests are too fragile
>  	* The output of the scapy tests is only visible in the opened
> terminal windows, so you can't see it when using --no-X, or (vitally)
> if the error causes the terminal to close.
> 
> --
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1] http://tldp.org/LDP/abs/html/exitcodes.html
> [2] https://github.com/niksu/pax/pull/12#issuecomment-235276667
> [3]
> https://github.com/notifications/unsubscribe-auth/ABDTTrqjDTI85fUHV-sTYoRnXP6YqpO4ks5qZhM2gaJpZM4JNj7N

-- 
http://www.cl.cam.ac.uk/~ns441/

niksu commented:
Thanks for looking into https://github.com/niksu/pax/pull/12#issuecomment-234261042, I changed the permissions as you suggested:
```
$ git status
...
	modified:   examples/Nat/nat_scapy_tests.py
	modified:   examples/Nat/nat_topo.py
	modified:   examples/pax_mininet_node.py
```
where the change was visible as
```
$ git diff
diff --git a/examples/Nat/nat_scapy_tests.py b/examples/Nat/nat_scapy_tests.py
old mode 100644
new mode 100755
diff --git a/examples/Nat/nat_topo.py b/examples/Nat/nat_topo.py
old mode 100644
new mode 100755
diff --git a/examples/pax_mininet_node.py b/examples/pax_mininet_node.py
old mode 100644
new mode 100755
```
and this seems to have fixed the issue, so it's worth committing this permission change to the repo. This is the output I got:

```
$ sudo ./examples/nat_topo.py --no-X test
sudo: ./examples/nat_topo.py: command not found
pal-pursuit1:pax$ sudo ./examples/Nat/nat_topo.py --no-X test
action: test, X_windows: False, hold_open: False
*** Creating network
*** Adding controller
*** Adding hosts:
in1 in2 nat0 out0 
*** Adding switches:

*** Adding links:
(nat0, in1) (nat0, in2) (nat0, out0) 
*** Configuring hosts
in1 in2 nat0 out0 
*** Starting controller
c0 
*** Starting 0 switches

Network started
Setting the gateway on in1 to nat0
Setting the gateway on in2 to nat0
Starting Pax NAT process on nat0:
  nat0> $ Bin/Pax.exe examples/Nat/nat_wiring_test.json examples/Bin/Examples.dll
Connecting from in1 to out0:
  out0> $ echo Hello, are you there? | netcat -l 12001 &
  in1> $ netcat -n 10.0.0.4 12001
  in1> RCV: 'Hello, are you there?'
Correct data received

Sending data via UDP from in1 to out0:
  out0> $ echo Yes, UDP | nc -u -l 12001 -q1
  in1> $ echo Hello, are you there UDP? | nc -u 10.0.0.4 12001 -q1
  out0> RCV: 'Hello, are you there UDP?'
  in1> RCV: 'Yes, UDP'
Correct data received on the server
Correct data received on the client

Scapy test #1
  out0> $ examples/Nat/nat_scapy_tests.py server 35001
  in1> $ sleep 1
  in1> $ examples/Nat/nat_scapy_tests.py client
  in1> $ echo $?
  out0> $ echo $?
Scapy test #1 passed

Scapy test #2
  out0> $ examples/Nat/nat_scapy_tests.py server2 35002
  in1> $ sleep 1
  in1> $ examples/Nat/nat_scapy_tests.py client2
  in1> $ echo $?
  out0> $ echo $?
Scapy test #2 passed
Show Pax output? (y/N)

*** Stopping 1 controllers
c0 
*** Stopping 3 links
...
*** Stopping 0 switches

*** Stopping 4 hosts
in1 in2 nat0 out0 
*** Done
```

niksu commented:
The problem described in https://github.com/niksu/pax/pull/12#issuecomment-234263211 still seems to be live, even with the permissions changed as mentioned in https://github.com/niksu/pax/pull/12#issuecomment-235276667 and https://github.com/niksu/pax/pull/12#issuecomment-235435849. Do you get similar behaviour, or is the shut-down tidy when you press `^D`?

```
$ sudo ./examples/Nat/nat_topo.py --no-X run
action: run, X_windows: False, hold_open: False
*** Creating network
*** Adding controller
*** Adding hosts:
in1 in2 nat0 out0 
*** Adding switches:

*** Adding links:
(nat0, in1) (nat0, in2) (nat0, out0) 
*** Configuring hosts
in1 in2 nat0 out0 
*** Starting controller
c0 
*** Starting 0 switches

Network started
Setting the gateway on in1 to nat0
Setting the gateway on in2 to nat0
*** Starting CLI:
mininet>
```

(then I press `^D`)

```
*** Stopping 1 controllers
c0 
*** Stopping 3 links
...
*** Stopping 0 switches

*** Stopping 4 hosts
in1 in2 nat0 out0 
*** Done
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "build/bdist.linux-x86_64/egg/mininet/cli.py", line 87, in <lambda>
    atexit.register( lambda: write_history_file( history_path ) )
IOError: [Errno 13] Permission denied
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "build/bdist.linux-x86_64/egg/mininet/cli.py", line 87, in <lambda>
IOError: [Errno 13] Permission denied
```

TMVector commented:
The shutdown is tidy when I run it. Perhaps try
```
sudo mn
```
and
```
sudo mn --custom examples/Nat/nat_topo.py --topo nat
```
and see if those error when you do `^D`.

niksu commented:
You're right, even `sudo mn` results in this issue on my setup. It seem to be a permission issue relating to `~/.mininet_history`. Thank you for looking into it.
Are there any outstanding issue? If not then we can merge this PR in.

TMVector commented:
I believe that all the issues are resolved. I have done a small [refactor](https://github.com/niksu/pax/pull/12/commits/6f6cddf958317e14e96ac04e5efa3977abb10815) of the TcpState class in an attempt to make the state transition clearer.

niksu commented:
That's superb! Many thanks for this great contribution and the sustained effort it required.
